nexus/src/db/subnet_allocation.rs

```diff
@@ -0,0 +1,347 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Diesel queries used for subnet allocation
+
+use crate::db;
+use crate::db::identity::Resource;
+use crate::db::model::IncompleteNetworkInterface;
+use chrono::{DateTime, Utc};
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::query_builder::*;
+use diesel::sql_types;
+use std::convert::TryFrom;
+use uuid::Uuid;
+
+/// Used for allocating an IP as part of [`NetworkInterface`] construction.
+///
+/// This is a query equivalent to:
+/// SELECT <id> AS id, <name> AS name, <description> AS description,
+///        <time_created> AS time_created, <time_modified> AS time_modified,
+///        <instance_id> AS instance_id, <vpc_id> AS vpc_id,
+///        <subnet_id> AS subnet_id, <mac> AS mac, <block_bASe> + off AS ip
+///   FROM
+///        generate_series(5, <num_addresses_in_block>) AS off
+///   LEFT OUTER JOIN
```

Comment:
This is pretty clever.  I have a few questions about how it works.

Do you have an idea about how this query behaves with large subnets?  I'm thinking of a few cases, supposing we have a subnet like 10.0.0.0/8 with 16M addresses:

- Does this always generate the whole list first?  So that even if the first address is free, we still generate all 16M?
- Does it search through each item in the list sequentially until it finds one that's free?  Does that mean as we fill up the subnet, it'll take longer to run?

Unrelatedly, I think this has the property that two concurrent allocations will attempt to allocate the same IP, which I think is going to be a problem when people use Terraform or CI or something else that provisions multiple Instances concurrently.

Even if the answers aren't great, that's not necessarily a blocker here and I don't have a better suggestion offhand.  I looked briefly when I was writing [this part of RFD 192](https://github.com/oxidecomputer/rfd/tree/master/rfd/0192#example-allocating-an-ip-address) but didn't find an obvious great solution.  But I think we'd want to come back to this if it doesn't scale that well.

Reply:
The query doesn't generate the whole list first (execution time doesn't seem to vary at all with the value of the end parameter, even if it's 2^63-1). It does take longer to execute the more full the space is.

It will have that problem with two concurrent runs (and then one of them failing the insert into network_interface due to unique violation). I couldn't figure out a way to do a simple allocation strategy that would avoid this.

Longer term, moving to something like a scheme where we store bitmaps of sub-blocks of addresses could be effective (so the query becomes a select for a bitmap that has any unset bits + a conditional update to allocate from it, retrying the whole process on failure). That select could potentially be multiple selects if we wanted it to be a tree structure. We'll want a similar allocator pattern for the MAC addresses. Using a SQL database might not be the best fit for these allocators, but does provide substantial simplicity for now

Reply:
That all sounds fine.  I'd just make sure we've recorded this as something to come back to.